File: crates/op-state-store/src/subid_ui.rs
//! Subid → UI presentation remap (render path only).
//!
//! Source of truth for the *rules*: `docs/subid-taxonomy.md` (dual id + seven
//! categories). This module does **not** invent taxonomy from a live scrape.
//!
//! Identity stays `uuid` + `subid`. Remap is strictly for rendering:
//! display / arrangement / priority / audience / element-type **role**.
//! Concrete json-render names (`Card`, `Button`, …) are out of scope here —
//! those resolve later in the projector against the gallery/catalog.

use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

/// The seven subid categories from `docs/subid-taxonomy.md`.
pub const SUBID_CATEGORIES: &[&str] = &["src", "prj", "sch", "mut", "obs", "evt", "exp"];

/// Presentation roles — data-semantic only (no json-render component names).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum UiRole {
    /// `exp.*` — consumer-facing surface
    Surface,
    /// `obs` scalar — read-only value
    DisplayValue,
    /// `obs` boolean — on/off state
    StateFlag,
    /// `obs` list of records
    CollectionView,
    /// `obs` record
    RecordView,
    /// `obs` list of scalars
    ValueList,
    /// `mut` boolean
    BinaryControl,
    /// `mut` string
    TextControl,
    /// `mut` integer / number
    NumericControl,
    /// `mut` list of scalars
    MultiChoice,
    /// `mut` list of records
    EditableCollection,
    /// `mut` record
    RecordEditor,
    /// `mut` otherwise / unstructured
    StructuredControl,
    /// `sch.*` — validation attaches to controls
    ValidationCarrier,
    /// `src.*` — hydration / ingress
    HydrationSource,
    /// `evt.*` — trigger / audit binding
    TriggerBinding,
    /// `prj.*` — projection / repeat binding
    RepeatBinding,
}

impl UiRole {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Surface => "surface",
            Self::DisplayValue => "display-value",
            Self::StateFlag => "state-flag",
            Self::CollectionView => "collection-view",
            Self::RecordView => "record-view",
            Self::ValueList => "value-list",
            Self::BinaryControl => "binary-control",
            Self::TextControl => "text-control",
            Self::NumericControl => "numeric-control",
            Self::MultiChoice => "multi-choice",
            Self::EditableCollection => "editable-collection",
            Self::RecordEditor => "record-editor",
            Self::StructuredControl => "structured-control",
            Self::ValidationCarrier => "validation-carrier",
            Self::HydrationSource => "hydration-source",
            Self::TriggerBinding => "trigger-binding",
            Self::RepeatBinding => "repeat-binding",
        }
    }
}

/// Normalized field shape used only to refine `obs`/`mut` roles.
/// Derived from schema field types when available; optional for method-only rows.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum UiFieldShape {
    Boolean,
    String,
    Integer,
    Float,
    ListScalar,
    ListRecord,
    Record,
    Any,
}

/// Extract the category (first segment) from a subid. Doc pattern:
/// `<category>.<component-type>.<subject>.<verb>[.<facet>][@vN]`
pub fn subid_category(subid: &str) -> Option<&str> {
    let base = subid.split('@').next().unwrap_or(subid);
    let cat = base.split('.').next()?;
    if SUBID_CATEGORIES.contains(&cat) {
        Some(cat)
    } else {
        None
    }
}

/// Doc-derived remap: category (+ optional shape) → presentation role.
///
/// Only the **category** drives the role. Remaining subid segments stay the
/// unique element join key; they are not reinterpreted here.
pub fn ui_role_from_subid(subid: &str, shape: Option<UiFieldShape>) -> Option<UiRole> {
    let cat = subid_category(subid)?;
    Some(match cat {
        "exp" => UiRole::Surface,
        "sch" => UiRole::ValidationCarrier,
        "src" => UiRole::HydrationSource,
        "evt" => UiRole::TriggerBinding,
        "prj" => UiRole::RepeatBinding,
        "obs" => match shape.unwrap_or(UiFieldShape::Any) {
            UiFieldShape::Boolean => UiRole::StateFlag,
            UiFieldShape::ListRecord => UiRole::CollectionView,
            UiFieldShape::Record => UiRole::RecordView,
            UiFieldShape::ListScalar => UiRole::ValueList,
            _ => UiRole::DisplayValue,
        },
        "mut" => match shape.unwrap_or(UiFieldShape::Any) {
            UiFieldShape::Boolean => UiRole::BinaryControl,
            UiFieldShape::String => UiRole::TextControl,
            UiFieldShape::Integer | UiFieldShape::Float => UiRole::NumericControl,
            UiFieldShape::ListScalar => UiRole::MultiChoice,
            UiFieldShape::ListRecord => UiRole::EditableCollection,
            UiFieldShape::Record => UiRole::RecordEditor,
            UiFieldShape::Any => UiRole::StructuredControl,
        },
        _ => return None,
    })
}

/// One populated catalog row for dump / API.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UiSubidProjection {
    pub plugin_id: String,
    pub kind: String,
    pub id: String,
    pub subid: String,
    pub category: String,
    pub role: String,
    /// Unique join key = subid segments after category (stable element id).
    pub element_key: String,
}

/// Strip category; remainder is the unique element key (plus `@vN` if present).
pub fn element_key_from_subid(subid: &str) -> String {
    let (base, ver) = match subid.split_once('@') {
        Some((b, v)) => (b, Some(v)),
        None => (subid, None),
    };
    let mut parts = base.split('.');
    let _cat = parts.next();
    let rest: Vec<&str> = parts.collect();
    let key = rest.join(".");
    match ver {
        Some(v) if !key.is_empty() => format!("{key}@{v}"),
        Some(v) => format!("@{v}"),
        None => key,
    }
}

/// Map a sealed `PluginSchema`'s fields + methods to presentation rows.
pub fn project_schema_ui(
    plugin_id: &str,
    schema: &crate::PluginSchema,
) -> Vec<UiSubidProjection> {
    let mut out = Vec::new();

    if let Some(sid) = schema.subids.get("__schema__") {
        if let Some(row) = row(plugin_id, "schema", plugin_id, sid, None) {
            out.push(row);
        }
    }

    for (fname, field) in &schema.fields {
        let Some(sid) = schema.subids.get(fname) else {
            continue;
        };
        let shape = Some(shape_from_field_type(&field.field_type));
        if let Some(row) = row(plugin_id, "field", fname, sid, shape) {
            out.push(row);
        }
    }

    for (mname, method) in &schema.methods {
        let sid = method.subid.as_str();
        if sid.is_empty() {
            continue;
        }
        let id = if method.name.is_empty() {
            mname.as_str()
        } else {
            method.name.as_str()
        };
        if let Some(row) = row(plugin_id, "method", id, sid, None) {
            out.push(row);
        }
    }

    out.sort_by(|a, b| {
        (&a.kind, &a.id, &a.subid).cmp(&(&b.kind, &b.id, &b.subid))
    });
    out
}

/// Role histogram for a dump (what populated).
pub fn role_population(rows: &[UiSubidProjection]) -> BTreeMap<String, usize> {
    let mut m = BTreeMap::new();
    for r in rows {
        *m.entry(r.role.clone()).or_insert(0) += 1;
    }
    m
}

fn row(
    plugin_id: &str,
    kind: &str,
    id: &str,
    subid: &str,
    shape: Option<UiFieldShape>,
) -> Option<UiSubidProjection> {
    let category = subid_category(subid)?.to_string();
    let role = ui_role_from_subid(subid, shape)?;
    Some(UiSubidProjection {
        plugin_id: plugin_id.to_string(),
        kind: kind.to_string(),
        id: id.to_string(),
        subid: subid.to_string(),
        category,
        role: role.as_str().to_string(),
        element_key: element_key_from_subid(subid),
    })
}

fn shape_from_field_type(ft: &crate::FieldType) -> UiFieldShape {
    use crate::FieldType;
    match ft {
        FieldType::Boolean => UiFieldShape::Boolean,
        FieldType::String | FieldType::Enum(_) => UiFieldShape::String,
        FieldType::Integer => UiFieldShape::Integer,
        FieldType::Float => UiFieldShape::Float,
        FieldType::Object(_) => UiFieldShape::Record,
        FieldType::Array(inner) => match inner.as_ref() {
            FieldType::Object(_) => UiFieldShape::ListRecord,
            FieldType::String
            | FieldType::Integer
            | FieldType::Float
            | FieldType::Boolean
            | FieldType::Enum(_) => UiFieldShape::ListScalar,
            _ => UiFieldShape::ListScalar,
        },
        FieldType::OneOf(_) | FieldType::Any => UiFieldShape::Any,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn category_only_from_doc_seven() {
        assert_eq!(subid_category("obs.software.plugin.xray.status@v1"), Some("obs"));
        assert_eq!(subid_category("mut.service.llm.generate@v1"), Some("mut"));
        assert_eq!(subid_category("nope.software.foo.bar@v1"), None);
    }

    #[test]
    fn roles_follow_category_not_component_names() {
        assert_eq!(
            ui_role_from_subid("exp.software.plugin.json-render.components@v1", None),
            Some(UiRole::Surface)
        );
        assert_eq!(
            ui_role_from_subid(
                "obs.software.plugin.cognitive-mcp.running@v1",
                Some(UiFieldShape::Boolean)
            ),
            Some(UiRole::StateFlag)
        );
        assert_eq!(
            ui_role_from_subid(
                "mut.service.blockchain.snapshot-interval@v1",
                Some(UiFieldShape::Integer)
            ),
            Some(UiRole::NumericControl)
        );
        // Never Card/Button — role string is abstract
        assert_eq!(UiRole::BinaryControl.as_str(), "binary-control");
    }

    #[test]
    fn element_key_strips_category_keeps_rest() {
        assert_eq!(
            element_key_from_subid("mut.service.llm.generate@v1"),
            "service.llm.generate@v1"
        );
    }
}

File: docs/architecture/zen-review-net-fabric/all-specs-archive/home-kiro/sessions/3799c34c17f6f960/sess_f2720a71-e73e-4d65-a928-66b68c4eab9d/snapshots/06bb1a81/srv/git/operation-dashboard-ui-07/src/lib/subid-ui.ts
/**
 * TypeScript mirror of `op-state-store/src/subid_ui.rs`.
 *
 * Types only — no logic. The Rust side is authoritative; this file exists so
 * the spec generator can work with the same type vocabulary.
 */

/**
 * The seven OSCAL subid categories from docs/subid-taxonomy.md.
 */
export const SUBID_CATEGORIES = ["src", "prj", "sch", "mut", "obs", "evt", "exp"] as const;
export type SubidCategory = (typeof SUBID_CATEGORIES)[number];

/**
 * Presentation roles — data-semantic only (no json-render component names).
 *
 * These are the kebab-case strings from `UiRole::as_str()` in Rust.
 */
export const UI_ROLES = [
  "surface",
  "display-value",
  "state-flag",
  "collection-view",
  "record-view",
  "value-list",
  "binary-control",
  "text-control",
  "numeric-control",
  "multi-choice",
  "editable-collection",
  "record-editor",
  "structured-control",
  "validation-carrier",
  "hydration-source",
  "trigger-binding",
  "repeat-binding",
] as const;

export type UiRole = (typeof UI_ROLES)[number];

/**
 * Normalized field shape used to refine obs/mut roles.
 * Derived from schema field types when available.
 */
export type UiFieldShape =
  | "boolean"
  | "string"
  | "integer"
  | "float"
  | "list-scalar"
  | "list-record"
  | "record"
  | "any";

/**
 * One populated catalog row for dump / API.
 * Mirrors `UiSubidProjection` in Rust.
 *
 * Returned by `GET /api/ui-model/plugin-schema/:plugin` in the `ui_projection` field.
 */
export interface UiSubidProjection {
  /** Plugin that owns this projection. */
  plugin_id: string;
  /** "field" | "method" | "schema" */
  kind: string;
  /** Field or method name. */
  id: string;
  /** Full subid string, e.g. "obs.software.plugin.xray.status@v1" */
  subid: string;
  /** Category extracted from subid (obs, mut, exp, ...) */
  category: SubidCategory | string;
  /** Presentation role (display-value, state-flag, ...) */
  role: UiRole;
  /** Unique join key = subid segments after category. */
  element_key: string;
}

/**
 * A plugin-owned UI surface route.
 * Mirrors `UiSurfaceRoute` in op-gallery-gen context.rs.
 */
export interface UiSurfaceRoute {
  /** Route path, e.g. "/antigravity/chat" */
  path: string;
  /** Human-readable name for nav */
  name?: string;
  /** Schema scope (which fields this surface shows) */
  schema?: string;
  /** Original raw object from the plugin */
  raw?: unknown;
}

/**
 * A plugin's ui_surfaces projection.
 * Mirrors `UiSurfaceProjection` in op-gallery-gen context.rs.
 */
export interface UiSurfaceProjection {
  /** Subids that declared ui_surfaces (for traceability) */
  subids: string[];
  /** Authoritative routes */
  routes: UiSurfaceRoute[];
  /** Where the routes came from ("default" | "example" | null) */
  value_source?: string;
}

/**
 * Check if a ui_surfaces projection is authoritative (has subids and routes).
 */
export function isAuthoritative(projection: UiSurfaceProjection | null | undefined): boolean {
  if (!projection) return false;
  return projection.subids.length > 0 && projection.routes.length > 0;
}

/**
 * Extract the category (first segment) from a subid.
 * Returns null if the category is not one of the seven known categories.
 */
export function subidCategory(subid: string): SubidCategory | null {
  const base = subid.split("@")[0] ?? subid;
  const cat = base.split(".")[0];
  if (cat && (SUBID_CATEGORIES as readonly string[]).includes(cat)) {
    return cat as SubidCategory;
  }
  return null;
}

/**
 * Derive a UiRole from a subid + optional field shape.
 * Mirrors `ui_role_from_subid` in Rust.
 */
export function uiRoleFromSubid(subid: string, shape?: UiFieldShape): UiRole | null {
  const cat = subidCategory(subid);
  if (!cat) return null;

  switch (cat) {
    case "exp":
      return "surface";
    case "sch":
      return "validation-carrier";
    case "src":
      return "hydration-source";
    case "evt":
      return "trigger-binding";
    case "prj":
      return "repeat-binding";
    case "obs":
      switch (shape) {
        case "boolean":
          return "state-flag";
        case "list-record":
          return "collection-view";
        case "record":
          return "record-view";
        case "list-scalar":
          return "value-list";
        default:
          return "display-value";
      }
    case "mut":
      switch (shape) {
        case "boolean":
          return "binary-control";
        case "string":
          return "text-control";
        case "integer":
        case "float":
          return "numeric-control";
        case "list-scalar":
          return "multi-choice";
        case "list-record":
          return "editable-collection";
        case "record":
          return "record-editor";
        default:
          return "structured-control";
      }
    default:
      return null;
  }
}

/**
 * Strip category from subid; remainder is the unique element key.
 * Mirrors `element_key_from_subid` in Rust.
 */
export function elementKeyFromSubid(subid: string): string {
  const [base, ver] = subid.split("@");
  const parts = (base ?? "").split(".");
  parts.shift(); // remove category
  const key = parts.join(".");
  if (ver && key) return `${key}@${ver}`;
  if (ver) return `@${ver}`;
  return key;
}
